src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs

```diff
@@ -580,7 +727,7 @@ private static IAggregator GetMinMaxAggType(IServiceProvider serviceProvider, IR
         // Average(1,2,3)
         public static FormulaValue Average(EvalVisitor runner, EvalVisitorContext context, IRContext irContext, FormulaValue[] args)
         {
-            return RunAggregator(new AverageAgg(), irContext, args);
+            return RunAggregator(irContext.ResultType == FormulaType.Decimal ? new AverageDecimalAgg() : new AverageAgg(), irContext, args);

```

Comment:
>irContext.ResultType == FormulaType.Decimal [](http://example.com/codeflow?start=33&length=43)

Design-wise, this kind of seems like a bit of a smell. We should likely have resolved this at IRTranslation time, making decisions this late based on type seems wrong.

+ @MikeStall

Reply:
McCall and I chatted and believe that we should move forward with the approach in the PR.  

We both agree that likely long term we should revise our function overload resolution logic, to better accommodate both Decimal and UDFs, but we don't have to do that now.  For Decimal, we can use the CheckTypes approach as currently coded, and UDFs can have their own rules as a new feature.  We have a discussion about UDFs scheduled for tomorrow.